core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -105,12 +105,46 @@
     this.overshadowedGroups = new TreeMap<>();
   }
 
-  OvershadowableManager(OvershadowableManager<T> other)
+  public OvershadowableManager<T> copyVisible()
   {
-    this.knownPartitionChunks = new HashMap<>(other.knownPartitionChunks);
-    this.standbyGroups = new TreeMap<>(other.standbyGroups);
-    this.visibleGroupPerRange = new TreeMap<>(other.visibleGroupPerRange);
-    this.overshadowedGroups = new TreeMap<>(other.overshadowedGroups);
+    final OvershadowableManager<T> copy = new OvershadowableManager<>();
+    visibleGroupPerRange.forEach((partitionRange, versionToGroups) -> {
+      // There should be only one group per partition range
+      final AtomicUpdateGroup<T> group = versionToGroups.values().iterator().next();
+      group.getChunks().forEach(chunk -> copy.knownPartitionChunks.put(chunk.getChunkNumber(), chunk));
+
+      copy.visibleGroupPerRange.put(
+          partitionRange,
+          new SingleEntryShort2ObjectSortedMap<>(group.getMinorVersion(), AtomicUpdateGroup.copy(group))
+      );
+    });
+    return copy;
+  }
+
+  public OvershadowableManager<T> deepCopy()
```

Comment:
should this override `public OvershadowableManager<T> clone()` instead?

Reply:
From the Item 13 (Override clone judiciously) of the book "Effective Java",

> Is all this complexity really necessary? Rarely. If you extend a class that already implements `Cloneable`, you have little choice but to implement a well-behaved `clone` method. Otherwise, you are usually better off providing an alternative means of object copying. **A better approach to object copying is to provide a _copy constructor_ or _copy factory_**. 

I changed this method to a static factory based on this.